storey/sources.py

```diff
@@ -622,49 +620,85 @@ def __init__(
+            if isinstance(self._key_field, list):
+                key = []
+                for key_field in self._key_field:
+                    key.append(
+                        self._get_by_field_or_index(field=key_field, body=body, field_type="key", raise_exception=True)
+                    )
+            else:
+                key = self._get_by_field_or_index(
+                    field=self._key_field, body=body, field_type="key", raise_exception=True
+                )
+        return key
+
+    def _get_element(self, body: OrderedDict):
+        return dict(body)
+
     async def _run_loop(self):
         for df in self._dfs:
             for namedtuple in df.itertuples():
-                create_event = True
-                body = namedtuple._asdict()
+                body = OrderedDict(namedtuple._asdict())
                 index = body.pop("Index")
                 if len(df.index.names) > 1:
                     for i, index_column in enumerate(df.index.names):
                         body[index_column] = index[i]
                 elif df.index.names[0] is not None:
                     body[df.index.names[0]] = index
-
-                key = None
-                if self._key_field:
-                    if isinstance(self._key_field, list):
-                        key = []
-                        for key_field in self._key_field:
-                            if key_field not in body or pandas.isna(body[key_field]):
-                                create_event = False
-                                break
-                            key.append(body[key_field])
-                    else:
-                        key = body[self._key_field]
-                        if key is None:
-                            create_event = False
-                if create_event:
+                try:
+                    key = self._get_key(body=body)
                     if self._id_field:
-                        id = body[self._id_field]
+                        line_id = self._get_by_field_or_index(
+                            field=self._id_field, body=body, field_type="id", raise_exception=False
+                        )
                     else:
-                        id = self._get_uuid()
-                    event = Event(body, key=key, id=id)
+                        line_id = self._get_uuid()
+                    element = self._get_element(body=body)
+                    event = Event(element, key=key, id=line_id)
                     await self._do_downstream(event)
-                else:
+                except self.NoneKeyException as key_error:
                     if self.context:
-                        self.context.logger.error(f"For {body} value of key {key_field} is None")
+                        self.context.logger.error(str(key_error))
+
         return await self._do_downstream(_termination_obj)
 
+    def _get_by_field_or_index(self, field, body: OrderedDict, field_type: str, raise_exception=False):
```

Comment:
"type" has a meaning that does not make sense here. Need a better name for this parameter.